File: services/order/src/main/kotlin/com/codejsha/bookstore/order/domain/model/OrderMapper.kt
package com.codejsha.bookstore.order.domain.model

import com.codejsha.bookstore.order.application.port.saga.PlaceOrderPayload
import com.codejsha.bookstore.service.application.port.openapi.model.OrderCreateWebReq
import com.codejsha.bookstore.service.application.port.openapi.model.OrderFindAllWebParam
import com.codejsha.bookstore.service.application.port.openapi.model.OrderUpdateWebReq
import com.codejsha.bookstore.service.application.port.pb.orderpb.OrderFindAllProtoReq

sealed class OrderMapper {
    companion object {
        fun toOrderDto(req: OrderCreateWebReq): OrderDto = req.toOrderDto()
        fun toOrderDto(id: Long, req: OrderUpdateWebReq): OrderDto = req.toOrderDto(id)
        fun toOrderDto(payload: PlaceOrderPayload): OrderDto = payload.toOrderDto()

        fun toFilterCondition(param: OrderFindAllWebParam): FilterCondition = param.toFilterCondition()
        fun toFilterCondition(req: OrderFindAllProtoReq): FilterCondition = req.toFilterCondition()
    }
}

private fun OrderCreateWebReq.toOrderDto(): OrderDto {
    return OrderDto(
        id = null,
        userId = this.userId,
        totalPrice = this.totalPrice,
        status = this.status,
        orderItems = this.orderItems?.map {
            OrderItemDto(
                bookId = requireNotNull(it.bookId),
                quantity = requireNotNull(it.quantity)
            )
        },
        paymentType = this.paymentType,
        cardNumber = this.cardNumber
    )
}

private fun OrderUpdateWebReq.toOrderDto(id: Long): OrderDto {
    return OrderDto(
        id = id,
        userId = this.userId,
        totalPrice = this.totalPrice,
        status = this.status,
        orderItems = this.orderItems?.map {
            OrderItemDto(
                bookId = requireNotNull(it.bookId),
                quantity = requireNotNull(it.quantity)
            )
        },
        paymentType = this.paymentType,
        cardNumber = this.cardNumber
    )
}

private fun PlaceOrderPayload.toOrderDto(): OrderDto {
    return OrderDto(
        id = null,
        userId = this.userId,
        totalPrice = this.totalPrice,
        status = this.status,
        orderItems = this.orderItems.map {
            OrderItemDto(
                bookId = requireNotNull(it.bookId),
                quantity = requireNotNull(it.quantity)
            )
        },
        paymentType = this.paymentType,
        cardNumber = this.cardNumber
    )
}

private fun OrderFindAllWebParam.toFilterCondition(): FilterCondition {
    return FilterCondition(
        userId = this.userId,
        filter = FilterCondition.QueryFilter(
            sort = this.sort,
            order = this.order,
            limit = this.limit,
            offset = this.offset
        )
    )
}

private fun OrderFindAllProtoReq.toFilterCondition(): FilterCondition {
    return FilterCondition(
        userId = this.userId,
        filter = FilterCondition.QueryFilter(
            sort = this.filter.sort,
            order = this.filter.order,
            limit = this.filter.limit,
            offset = this.filter.offset
        )
    )
}

File: services/customer/internal/domain/model/mapper/order_mapper.go
package mapper

import (
	"fmt"

	"github.com/codejsha/bookstore-microservices/customer/internal/application/port/openapi"
	"github.com/codejsha/bookstore-microservices/customer/internal/application/port/pb/orderpb"
)

func NewOrderMapper() *OrderMapper {
	return &OrderMapper{}
}

type OrderMapper struct{}

func (m OrderMapper) ToOrderFindAllProtoReq(value interface{}) (*orderpb.OrderFindAllProtoReq, error) {
	switch value.(type) {
	case openapi.OrderFindAllWebParam:
		v := value.(openapi.OrderFindAllWebParam)
		status, err := orderStatusFromValue(v.Status)
		if err != nil {
			return nil, fmt.Errorf("failed to convert request to OrderFindAllRequest: %v", value)
		}
		request := &orderpb.OrderFindAllProtoReq{
			UserId: &v.Id,
			Status: &status,
			Filter: &orderpb.QueryFilter{
				Sort:   v.Sort,
				Order:  v.Order,
				Limit:  v.Limit,
				Offset: v.Offset,
			},
		}
		return request, nil
	default:
		return nil, fmt.Errorf("failed to convert request to OrderFindAllRequest: %v", value)
	}
}

func (m OrderMapper) ToOrderFindProtoReq(value interface{}) (*orderpb.OrderFindProtoReq, error) {
	switch value.(type) {
	case openapi.OrderFindWebParam:
		v := value.(openapi.OrderFindWebParam)
		request := &orderpb.OrderFindProtoReq{
			Id: v.OrderId,
		}
		return request, nil
	default:
		return nil, fmt.Errorf("failed to convert request to OrderFindRequest: %v", value)
	}
}

func orderStatusFromValue(value interface{}) (orderpb.OrderStatus, error) {
	switch value.(type) {
	case openapi.OrderStatus:
		v := value.(openapi.OrderStatus)
		switch v {
		case openapi.ORDERSTATUS_UNKNOWN:
			return orderpb.OrderStatus_UNKNOWN, nil
		case openapi.ORDERSTATUS_PENDING:
			return orderpb.OrderStatus_PENDING, nil
		case openapi.ORDERSTATUS_SHIPPING:
			return orderpb.OrderStatus_SHIPPING, nil
		case openapi.ORDERSTATUS_COMPLETED:
			return orderpb.OrderStatus_COMPLETED, nil
		case openapi.ORDERSTATUS_CANCELLED:
			return orderpb.OrderStatus_CANCELLED, nil
		default:
			return orderpb.OrderStatus_UNKNOWN, fmt.Errorf("failed to convert value to OrderStatus: %v", value)
		}
	default:
		return orderpb.OrderStatus_UNKNOWN, fmt.Errorf("failed to convert value to OrderStatus: %v", value)
	}
}
